colexec: some optimizations

yuzefovich opened the pull request:
**colexec: remove one of the Go maps from hash aggregator**

This commit switches usage of `map` to iteration over `[]uint64` when
building selection vectors in the hash aggregator. This is a lot more
efficient when group sizes are relatively large with moderate hit when
group sizes are small. This hit is reduced in a follow-up commit.

Release note: None

**colexec: more improvements to hash aggregator**

This commit removes the buffering stage of the hash aggregator as well
as removes the "append only" scratch batch that we're currently using.
The removal of buffering stage allows us to have smaller buffers without
sacrificing the performance. The removal of the scratch batch allows to
avoid copying over the data from the input batch and using that input
batch directly. We will be descructively modifying the selection vector
on that batch, but such behavior is acceptable because hash aggregator
owns the output batch, and the input batch will not be propagated
further.

This commit also bumps `hashAggFuncsAllocSize` from 16 to 64 which
gives us minor performance improvement in case of small group sizes.

Release note: None

**colexec: remove some allocations**

In a recent PR (for logical types plumbing) I introduced some
unnecessary allocations for unhandled type case - by taking a pointer
from a value in `[]types.T` slice. This commit fixes that.

Release note: None

yuzefovich committed:
colexec: remove one of the Go maps from hash aggregator

This commit switches usage of `map` to iteration over `[]uint64` when
building selection vectors in the hash aggregator. This is a lot more
efficient when group sizes are relatively large with moderate hit when
group sizes are small. This hit is reduced in a follow-up commit.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47942)
<!-- Reviewable:end -->

yuzefovich committed:
colexec: remove some allocations

In a recent PR (for logical types plumbing) I introduced some
unnecessary allocations for unhandled type case - by taking a pointer
from a value in `[]types.T` slice. This commit fixes that.

Release note: None

yuzefovich commented:
Second commit removes the allocations I mistakenly introduced:
```
name                                       old alloc/op   new alloc/op   delta
MergeJoiner/rows=1024-16                     1.58kB ± 0%    1.26kB ± 0%   ~     (p=0.100 n=3+3)
MergeJoiner/rows=4096-16                     7.27kB ± 0%    5.05kB ± 0%   ~     (p=0.100 n=3+3)
MergeJoiner/rows=16384-16                    30.1kB ± 0%    20.2kB ± 0%   ~     (p=0.100 n=3+3)
MergeJoiner/rows=1048576-16                  1.95MB ± 0%    1.29MB ± 0%   ~     (p=0.100 n=3+3)
MergeJoiner/oneSideRepeat-rows=1024-16       1.58kB ± 0%    1.26kB ± 0%   ~     (p=0.100 n=3+3)
MergeJoiner/oneSideRepeat-rows=4096-16       7.27kB ± 0%    5.04kB ± 0%   ~     (p=0.100 n=3+3)
MergeJoiner/oneSideRepeat-rows=16384-16      30.0kB ± 0%    20.1kB ± 0%   ~     (p=0.100 n=3+3)
MergeJoiner/oneSideRepeat-rows=1048576-16    1.94MB ± 0%    1.29MB ± 0%   ~     (p=0.100 n=3+3)
MergeJoiner/bothSidesRepeat-rows=1024-16     1.58kB ± 0%    1.26kB ± 0%   ~     (p=0.100 n=3+3)
MergeJoiner/bothSidesRepeat-rows=4096-16     7.50kB ± 0%    5.21kB ± 0%   ~     (p=0.100 n=3+3)
MergeJoiner/bothSidesRepeat-rows=16384-16    31.2kB ± 0%    21.0kB ± 0%   ~     (p=0.100 n=3+3)
MergeJoiner/bothSidesRepeat-rows=32768-16    63.2kB ± 0%    42.8kB ± 0%   ~     (p=0.100 n=3+3)
```
Compare it to the benchmarks of logical plumbing PR from [here](https://gist.github.com/yuzefovich/a4743518fb73c280464401ffdfc12751):
```
MergeJoiner/oneSideRepeat-rows=1024-24                                                               1.22kB ± 0%     1.61kB ± 0%   +31.48%  (p=0.008 n=5+5)
MergeJoiner/bothSidesRepeat-rows=1024-24                                                             1.22kB ± 0%     1.61kB ± 0%   +31.48%  (p=0.008 n=5+5)
MergeJoiner/bothSidesRepeat-rows=4096-24                                                             5.14kB ± 0%     7.64kB ± 0%   +48.59%  (p=0.008 n=5+5)
MergeJoiner/bothSidesRepeat-rows=32768-24                                                            43.3kB ± 0%     65.5kB ± 0%   +51.45%  (p=0.008 n=5+5)
MergeJoiner/oneSideRepeat-rows=4096-24                                                               4.85kB ± 0%     7.35kB ± 0%   +51.60%  (p=0.008 n=5+5)
MergeJoiner/rows=4096-24                                                                             4.84kB ± 0%     7.35kB ± 0%   +51.65%  (p=0.008 n=5+5)
MergeJoiner/bothSidesRepeat-rows=16384-24                                                            20.9kB ± 0%     31.9kB ± 0%   +52.41%  (p=0.016 n=5+4)
MergeJoiner/oneSideRepeat-rows=16384-24                                                              19.4kB ± 0%     30.3kB ± 0%   +56.60%  (p=0.008 n=5+5)
MergeJoiner/rows=16384-24                                                                            19.4kB ± 0%     30.3kB ± 0%   +56.64%  (p=0.008 n=5+5)
MergeJoiner/oneSideRepeat-rows=1048576-24                                                            1.24MB ± 0%     1.96MB ± 0%   +58.27%  (p=0.008 n=5+5)
MergeJoiner/rows=1048576-24                                                                          1.24MB ± 0%     1.96MB ± 0%   +58.28%  (p=0.008 n=5+5)
```
Thanks @jordanlewis

yuzefovich commented:
The benchmarks of the hash aggregator:
```
name                                                                             old speed      new speed       delta
Aggregator/MIN/hash/decimal/groupSize=1/hasNulls=false/numInputBatches=64-24     4.00MB/s ± 3%   3.08MB/s ± 2%  -22.87%  (p=0.000 n=10+10)
Aggregator/MIN/hash/decimal/groupSize=1/hasNulls=true/numInputBatches=64-24      3.79MB/s ± 2%   2.99MB/s ± 2%  -21.03%  (p=0.000 n=10+10)
Aggregator/MIN/hash/decimal/groupSize=4/hasNulls=false/numInputBatches=64-24     13.8MB/s ± 1%   12.1MB/s ± 1%  -12.55%  (p=0.000 n=10+10)
Aggregator/MIN/hash/decimal/groupSize=4/hasNulls=true/numInputBatches=64-24      13.7MB/s ± 1%   12.0MB/s ± 2%  -12.55%  (p=0.000 n=10+10)
Aggregator/MIN/hash/decimal/groupSize=16/hasNulls=false/numInputBatches=64-24    36.8MB/s ± 1%   37.6MB/s ± 0%   +2.14%  (p=0.000 n=10+10)
Aggregator/MIN/hash/decimal/groupSize=16/hasNulls=true/numInputBatches=64-24     36.5MB/s ± 1%   37.2MB/s ± 1%   +2.03%  (p=0.000 n=10+10)
Aggregator/MIN/hash/decimal/groupSize=64/hasNulls=false/numInputBatches=64-24    63.2MB/s ± 2%   79.7MB/s ± 1%  +26.14%  (p=0.000 n=10+10)
Aggregator/MIN/hash/decimal/groupSize=64/hasNulls=true/numInputBatches=64-24     62.4MB/s ± 2%   78.3MB/s ± 1%  +25.62%  (p=0.000 n=10+10)
Aggregator/MIN/hash/decimal/groupSize=256/hasNulls=false/numInputBatches=64-24   88.0MB/s ± 1%  112.9MB/s ± 1%  +28.23%  (p=0.000 n=10+10)
Aggregator/MIN/hash/decimal/groupSize=256/hasNulls=true/numInputBatches=64-24    86.4MB/s ± 0%  109.2MB/s ± 1%  +26.41%  (p=0.000 n=10+10)
Aggregator/MIN/hash/decimal/groupSize=1024/hasNulls=false/numInputBatches=64-24   101MB/s ± 1%    125MB/s ± 1%  +24.13%  (p=0.000 n=10+10)
Aggregator/MIN/hash/decimal/groupSize=1024/hasNulls=true/numInputBatches=64-24   98.5MB/s ± 0%  121.3MB/s ± 1%  +23.15%  (p=0.000 n=10+10)
```
Note that I have changed the group sizes, and it looks like there is a performance hit for groups smaller than 8 or so tuples and performance improvement when we go bigger. Also note that that this benchmark is using decimals, and on ints the performance gains are significantly higher in large group size cases.

yuzefovich commented:
Here are the benchmarks of two commits with more group sizes on ints:
```
name                                                                         old speed      new speed       delta
Aggregator/MIN/hash/int/groupSize=1/hasNulls=false/numInputBatches=64-24     4.24MB/s ± 2%   3.39MB/s ± 2%  -20.00%  (p=0.000 n=10+10)
Aggregator/MIN/hash/int/groupSize=1/hasNulls=true/numInputBatches=64-24      4.06MB/s ± 1%   3.30MB/s ± 0%  -18.52%  (p=0.000 n=10+10)
Aggregator/MIN/hash/int/groupSize=2/hasNulls=false/numInputBatches=64-24     7.90MB/s ± 1%   6.87MB/s ± 1%  -13.02%  (p=0.000 n=10+10)
Aggregator/MIN/hash/int/groupSize=2/hasNulls=true/numInputBatches=64-24      7.46MB/s ± 2%   6.57MB/s ± 2%  -11.99%  (p=0.000 n=10+10)
Aggregator/MIN/hash/int/groupSize=4/hasNulls=false/numInputBatches=64-24     15.5MB/s ± 1%   13.9MB/s ± 1%  -10.79%  (p=0.000 n=10+10)
Aggregator/MIN/hash/int/groupSize=4/hasNulls=true/numInputBatches=64-24      15.2MB/s ± 1%   13.8MB/s ± 1%   -9.73%  (p=0.000 n=10+10)
Aggregator/MIN/hash/int/groupSize=8/hasNulls=false/numInputBatches=64-24     28.8MB/s ± 1%   27.3MB/s ± 1%   -5.10%  (p=0.000 n=9+10)
Aggregator/MIN/hash/int/groupSize=8/hasNulls=true/numInputBatches=64-24      28.0MB/s ± 1%   26.7MB/s ± 0%   -4.38%  (p=0.000 n=10+9)
Aggregator/MIN/hash/int/groupSize=16/hasNulls=false/numInputBatches=64-24    49.1MB/s ± 1%   51.0MB/s ± 1%   +3.87%  (p=0.000 n=10+10)
Aggregator/MIN/hash/int/groupSize=16/hasNulls=true/numInputBatches=64-24     47.0MB/s ± 1%   49.1MB/s ± 0%   +4.42%  (p=0.000 n=10+8)
Aggregator/MIN/hash/int/groupSize=32/hasNulls=false/numInputBatches=64-24    75.2MB/s ± 1%   93.3MB/s ± 1%  +24.12%  (p=0.000 n=10+10)
Aggregator/MIN/hash/int/groupSize=32/hasNulls=true/numInputBatches=64-24     70.4MB/s ± 1%   87.7MB/s ± 0%  +24.48%  (p=0.000 n=10+10)
Aggregator/MIN/hash/int/groupSize=64/hasNulls=false/numInputBatches=64-24     102MB/s ± 2%    158MB/s ± 1%  +53.96%  (p=0.000 n=10+10)
Aggregator/MIN/hash/int/groupSize=64/hasNulls=true/numInputBatches=64-24     97.1MB/s ± 2%  140.8MB/s ± 1%  +45.04%  (p=0.000 n=9+10)
Aggregator/MIN/hash/int/groupSize=128/hasNulls=false/numInputBatches=64-24    134MB/s ± 0%    242MB/s ± 0%  +80.24%  (p=0.000 n=8+10)
Aggregator/MIN/hash/int/groupSize=128/hasNulls=true/numInputBatches=64-24     120MB/s ± 2%    205MB/s ± 0%  +70.89%  (p=0.000 n=9+10)
Aggregator/MIN/hash/int/groupSize=256/hasNulls=false/numInputBatches=64-24    180MB/s ± 0%    319MB/s ± 0%  +77.89%  (p=0.000 n=10+8)
Aggregator/MIN/hash/int/groupSize=256/hasNulls=true/numInputBatches=64-24     157MB/s ± 0%    255MB/s ± 1%  +62.31%  (p=0.000 n=10+10)
Aggregator/MIN/hash/int/groupSize=512/hasNulls=false/numInputBatches=64-24    207MB/s ± 1%    382MB/s ± 0%  +84.37%  (p=0.000 n=10+10)
Aggregator/MIN/hash/int/groupSize=512/hasNulls=true/numInputBatches=64-24     179MB/s ± 0%    295MB/s ± 0%  +64.84%  (p=0.000 n=10+10)
Aggregator/MIN/hash/int/groupSize=1024/hasNulls=false/numInputBatches=64-24   233MB/s ± 0%    427MB/s ± 0%  +83.10%  (p=0.000 n=10+9)
Aggregator/MIN/hash/int/groupSize=1024/hasNulls=true/numInputBatches=64-24    197MB/s ± 0%    321MB/s ± 0%  +62.51%  (p=0.000 n=10+10)
```

jordanlewis commented:
Looks good. I don't think we should forgo the small group size stuff, but I understand that with the current algorithm it would be hard to get small groups right. We need to think more about this though.

yuzefovich commented:
I have some more improvements coming up :) (nothing drastic though)

yuzefovich commented:
Actually, not too shabby:
```
Aggregator/MIN/hash/int/groupSize=1/hasNulls=false/numInputBatches=64-16     6.25MB/s ± 7%   8.35MB/s ± 1%   ~     (p=0.100 n=3+3)
Aggregator/MIN/hash/int/groupSize=1/hasNulls=true/numInputBatches=64-16      5.98MB/s ± 3%   7.98MB/s ± 1%   ~     (p=0.100 n=3+3)
Aggregator/MIN/hash/int/groupSize=2/hasNulls=false/numInputBatches=64-16     13.7MB/s ± 4%   17.4MB/s ± 2%   ~     (p=0.100 n=3+3)
Aggregator/MIN/hash/int/groupSize=2/hasNulls=true/numInputBatches=64-16      14.3MB/s ± 0%   15.6MB/s ± 4%   ~     (p=0.100 n=3+3)
Aggregator/MIN/hash/int/groupSize=4/hasNulls=false/numInputBatches=64-16     28.5MB/s ± 5%   32.7MB/s ± 9%   ~     (p=0.100 n=3+3)
Aggregator/MIN/hash/int/groupSize=4/hasNulls=true/numInputBatches=64-16      26.1MB/s ± 3%   35.5MB/s ± 3%   ~     (p=0.100 n=3+3)
Aggregator/MIN/hash/int/groupSize=8/hasNulls=false/numInputBatches=64-16     52.8MB/s ± 4%   64.1MB/s ± 4%   ~     (p=0.100 n=3+3)
Aggregator/MIN/hash/int/groupSize=8/hasNulls=true/numInputBatches=64-16      52.8MB/s ± 1%   64.3MB/s ± 2%   ~     (p=0.100 n=3+3)
Aggregator/MIN/hash/int/groupSize=16/hasNulls=false/numInputBatches=64-16     101MB/s ± 6%    122MB/s ± 1%   ~     (p=0.100 n=3+3)
Aggregator/MIN/hash/int/groupSize=16/hasNulls=true/numInputBatches=64-16     98.6MB/s ± 1%  118.3MB/s ± 3%   ~     (p=0.100 n=3+3)
Aggregator/MIN/hash/int/groupSize=32/hasNulls=false/numInputBatches=64-16     186MB/s ± 0%    206MB/s ± 3%   ~     (p=0.100 n=3+3)
Aggregator/MIN/hash/int/groupSize=32/hasNulls=true/numInputBatches=64-16      171MB/s ± 0%    207MB/s ± 1%   ~     (p=0.100 n=3+3)
Aggregator/MIN/hash/int/groupSize=64/hasNulls=false/numInputBatches=64-16     303MB/s ± 1%    351MB/s ± 2%   ~     (p=0.100 n=3+3)
Aggregator/MIN/hash/int/groupSize=64/hasNulls=true/numInputBatches=64-16      261MB/s ± 2%    335MB/s ± 0%   ~     (p=0.100 n=3+3)
Aggregator/MIN/hash/int/groupSize=128/hasNulls=false/numInputBatches=64-16    414MB/s ± 1%    487MB/s ± 1%   ~     (p=0.100 n=3+3)
Aggregator/MIN/hash/int/groupSize=128/hasNulls=true/numInputBatches=64-16     342MB/s ± 0%    463MB/s ± 2%   ~     (p=0.100 n=3+3)
Aggregator/MIN/hash/int/groupSize=256/hasNulls=false/numInputBatches=64-16    559MB/s ± 2%    583MB/s ± 1%   ~     (p=0.100 n=3+3)
Aggregator/MIN/hash/int/groupSize=256/hasNulls=true/numInputBatches=64-16     449MB/s ± 0%    558MB/s ± 2%   ~     (p=0.100 n=3+3)
Aggregator/MIN/hash/int/groupSize=512/hasNulls=false/numInputBatches=64-16    656MB/s ± 0%    674MB/s ± 0%   ~     (p=0.100 n=3+3)
Aggregator/MIN/hash/int/groupSize=512/hasNulls=true/numInputBatches=64-16     500MB/s ± 0%    642MB/s ± 2%   ~     (p=0.100 n=3+3)
Aggregator/MIN/hash/int/groupSize=1024/hasNulls=false/numInputBatches=64-16   705MB/s ± 1%    741MB/s ± 0%   ~     (p=0.100 n=3+3)
Aggregator/MIN/hash/int/groupSize=1024/hasNulls=true/numInputBatches=64-16    528MB/s ± 1%    696MB/s ± 0%   ~     (p=0.100 n=3+3)
```

yuzefovich committed:
colexec: more improvements to hash aggregator

This commit removes the buffering stage of the hash aggregator as well
as removes the "append only" scratch batch that we're currently using.
The removal of buffering stage allows us to have smaller buffers without
sacrificing the performance. The removal of the scratch batch allows to
avoid copying over the data from the input batch and using that input
batch directly. We will be descructively modifying the selection vector
on that batch, but such behavior is acceptable because hash aggregator
owns the output batch, and the input batch will not be propagated
further.

This commit also bumps `hashAggFuncsAllocSize` from 16 to 64 which
gives us minor performance improvement in case of small group sizes.

Release note: None

yuzefovich commented:
I removed the comments that were describing the benchmarks of my WIP.

Some observations from those comments:
- the first commit on its introduces about 20% performance hit with small group sizes but gives at least 20% performance improvement on large group sizes
- that gain can be up to 80% (when aggregating integers)
- the first commit reduces allocations in all cases.

The second commit removes the buffering stage which reduces the maximum length of `hashBuffer` to `coldata.BatchSize` which makes the approach of the first commit suffer less with small group sizes. As a result, we see gains across the board (full output is [here](https://gist.github.com/yuzefovich/fdc17dd109e5f40bd25b3e9f1802ba04)):
```
name                                                                             old speed      new speed       delta
Aggregator/SUM/hash/int/groupSize=1/hasNulls=false/numInputBatches=64-24         4.48MB/s ± 2%   5.23MB/s ± 1%   +16.70%  (p=0.000 n=10+10)
Aggregator/SUM/hash/int/groupSize=1/hasNulls=true/numInputBatches=64-24          4.18MB/s ± 2%   4.86MB/s ± 1%   +16.41%  (p=0.000 n=10+10)
Aggregator/SUM/hash/int/groupSize=2/hasNulls=false/numInputBatches=64-24         8.18MB/s ± 3%  10.66MB/s ± 5%   +30.32%  (p=0.000 n=10+10)
Aggregator/SUM/hash/int/groupSize=2/hasNulls=true/numInputBatches=64-24          7.77MB/s ± 1%  10.14MB/s ± 3%   +30.45%  (p=0.000 n=10+10)
Aggregator/SUM/hash/int/groupSize=4/hasNulls=false/numInputBatches=64-24         17.0MB/s ± 3%   21.7MB/s ± 1%   +27.66%  (p=0.000 n=10+10)
Aggregator/SUM/hash/int/groupSize=4/hasNulls=true/numInputBatches=64-24          16.3MB/s ± 3%   21.3MB/s ± 2%   +30.48%  (p=0.000 n=10+10)
Aggregator/SUM/hash/int/groupSize=8/hasNulls=false/numInputBatches=64-24         30.8MB/s ± 1%   42.0MB/s ± 1%   +36.29%  (p=0.000 n=9+10)
Aggregator/SUM/hash/int/groupSize=8/hasNulls=true/numInputBatches=64-24          29.8MB/s ± 1%   41.1MB/s ± 1%   +38.14%  (p=0.000 n=9+10)
Aggregator/SUM/hash/int/groupSize=16/hasNulls=false/numInputBatches=64-24        52.6MB/s ± 1%   82.2MB/s ± 1%   +56.40%  (p=0.000 n=10+8)
Aggregator/SUM/hash/int/groupSize=16/hasNulls=true/numInputBatches=64-24         50.1MB/s ± 1%   78.0MB/s ± 4%   +55.72%  (p=0.000 n=10+10)
Aggregator/SUM/hash/int/groupSize=32/hasNulls=false/numInputBatches=64-24        79.8MB/s ± 2%  143.7MB/s ± 2%   +79.98%  (p=0.000 n=10+10)
Aggregator/SUM/hash/int/groupSize=32/hasNulls=true/numInputBatches=64-24         75.7MB/s ± 4%  135.5MB/s ± 1%   +78.99%  (p=0.000 n=10+10)
Aggregator/SUM/hash/int/groupSize=64/hasNulls=false/numInputBatches=64-24         107MB/s ± 2%    226MB/s ± 0%  +111.44%  (p=0.000 n=9+10)
Aggregator/SUM/hash/int/groupSize=64/hasNulls=true/numInputBatches=64-24         99.2MB/s ± 3%  207.8MB/s ± 1%  +109.44%  (p=0.000 n=10+10)
Aggregator/SUM/hash/int/groupSize=128/hasNulls=false/numInputBatches=64-24        134MB/s ± 5%    320MB/s ± 0%  +139.87%  (p=0.000 n=10+10)
Aggregator/SUM/hash/int/groupSize=128/hasNulls=true/numInputBatches=64-24         123MB/s ± 6%    285MB/s ± 2%  +131.99%  (p=0.000 n=10+10)
Aggregator/SUM/hash/int/groupSize=256/hasNulls=false/numInputBatches=64-24        182MB/s ± 2%    403MB/s ± 0%  +121.56%  (p=0.000 n=10+10)
Aggregator/SUM/hash/int/groupSize=256/hasNulls=true/numInputBatches=64-24         159MB/s ± 1%    350MB/s ± 0%  +120.66%  (p=0.000 n=10+10)
Aggregator/SUM/hash/int/groupSize=512/hasNulls=false/numInputBatches=64-24        208MB/s ± 1%    457MB/s ± 0%  +119.58%  (p=0.000 n=10+8)
Aggregator/SUM/hash/int/groupSize=512/hasNulls=true/numInputBatches=64-24         182MB/s ± 0%    390MB/s ± 0%  +114.08%  (p=0.000 n=10+10)
Aggregator/SUM/hash/int/groupSize=1024/hasNulls=false/numInputBatches=64-24       234MB/s ± 0%    486MB/s ± 0%  +108.29%  (p=0.000 n=9+10)
Aggregator/SUM/hash/int/groupSize=1024/hasNulls=true/numInputBatches=64-24        202MB/s ± 1%    413MB/s ± 0%  +104.29%  (p=0.000 n=10+8)
Aggregator/SUM/hash/int/groupSize=2048/hasNulls=false/numInputBatches=64-24       274MB/s ± 1%    501MB/s ± 1%   +82.55%  (p=0.000 n=10+10)
Aggregator/SUM/hash/int/groupSize=2048/hasNulls=true/numInputBatches=64-24        230MB/s ± 1%    421MB/s ± 0%   +82.87%  (p=0.000 n=9+9)
Aggregator/SUM/hash/int/groupSize=4096/hasNulls=false/numInputBatches=64-24       278MB/s ± 1%    505MB/s ± 0%   +81.65%  (p=0.000 n=10+10)
Aggregator/SUM/hash/int/groupSize=4096/hasNulls=true/numInputBatches=64-24        233MB/s ± 0%    426MB/s ± 0%   +82.68%  (p=0.000 n=8+10)
Aggregator/SUM/hash/decimal/groupSize=1/hasNulls=false/numInputBatches=64-24     3.73MB/s ± 3%   4.21MB/s ± 2%   +12.96%  (p=0.000 n=10+10)
Aggregator/SUM/hash/decimal/groupSize=1/hasNulls=true/numInputBatches=64-24      3.60MB/s ± 1%   4.12MB/s ± 1%   +14.45%  (p=0.000 n=10+10)
Aggregator/SUM/hash/decimal/groupSize=2/hasNulls=false/numInputBatches=64-24     6.18MB/s ± 4%   7.22MB/s ± 1%   +16.95%  (p=0.000 n=10+10)
Aggregator/SUM/hash/decimal/groupSize=2/hasNulls=true/numInputBatches=64-24      6.09MB/s ± 3%   7.40MB/s ± 3%   +21.48%  (p=0.000 n=10+10)
Aggregator/SUM/hash/decimal/groupSize=4/hasNulls=false/numInputBatches=64-24     9.83MB/s ± 1%  11.89MB/s ± 2%   +20.97%  (p=0.000 n=10+10)
Aggregator/SUM/hash/decimal/groupSize=4/hasNulls=true/numInputBatches=64-24      10.1MB/s ± 1%   12.2MB/s ± 1%   +21.45%  (p=0.000 n=10+10)
Aggregator/SUM/hash/decimal/groupSize=8/hasNulls=false/numInputBatches=64-24     14.4MB/s ± 1%   16.8MB/s ± 2%   +16.63%  (p=0.000 n=10+9)
Aggregator/SUM/hash/decimal/groupSize=8/hasNulls=true/numInputBatches=64-24      15.0MB/s ± 1%   17.4MB/s ± 1%   +16.13%  (p=0.000 n=9+10)
Aggregator/SUM/hash/decimal/groupSize=16/hasNulls=false/numInputBatches=64-24    18.7MB/s ± 1%   21.7MB/s ± 2%   +15.99%  (p=0.000 n=10+10)
Aggregator/SUM/hash/decimal/groupSize=16/hasNulls=true/numInputBatches=64-24     19.7MB/s ± 1%   23.3MB/s ± 0%   +18.42%  (p=0.000 n=10+9)
Aggregator/SUM/hash/decimal/groupSize=32/hasNulls=false/numInputBatches=64-24    22.0MB/s ± 2%   26.1MB/s ± 0%   +18.61%  (p=0.000 n=10+10)
Aggregator/SUM/hash/decimal/groupSize=32/hasNulls=true/numInputBatches=64-24     23.0MB/s ± 1%   28.2MB/s ± 1%   +22.51%  (p=0.000 n=9+10)
Aggregator/SUM/hash/decimal/groupSize=64/hasNulls=false/numInputBatches=64-24    24.0MB/s ± 1%   28.7MB/s ± 1%   +19.54%  (p=0.000 n=10+10)
Aggregator/SUM/hash/decimal/groupSize=64/hasNulls=true/numInputBatches=64-24     25.3MB/s ± 1%   30.6MB/s ± 1%   +21.05%  (p=0.000 n=10+10)
Aggregator/SUM/hash/decimal/groupSize=128/hasNulls=false/numInputBatches=64-24   25.2MB/s ± 1%   30.1MB/s ± 1%   +19.48%  (p=0.000 n=9+10)
Aggregator/SUM/hash/decimal/groupSize=128/hasNulls=true/numInputBatches=64-24    26.7MB/s ± 1%   32.6MB/s ± 1%   +21.77%  (p=0.000 n=10+10)
Aggregator/SUM/hash/decimal/groupSize=256/hasNulls=false/numInputBatches=64-24   26.9MB/s ± 1%   30.8MB/s ± 1%   +14.68%  (p=0.000 n=10+10)
Aggregator/SUM/hash/decimal/groupSize=256/hasNulls=true/numInputBatches=64-24    28.7MB/s ± 1%   33.6MB/s ± 1%   +17.12%  (p=0.000 n=10+10)
Aggregator/SUM/hash/decimal/groupSize=512/hasNulls=false/numInputBatches=64-24   27.7MB/s ± 1%   31.5MB/s ± 1%   +13.84%  (p=0.000 n=10+10)
Aggregator/SUM/hash/decimal/groupSize=512/hasNulls=true/numInputBatches=64-24    29.3MB/s ± 1%   34.0MB/s ± 1%   +16.01%  (p=0.000 n=10+10)
Aggregator/SUM/hash/decimal/groupSize=1024/hasNulls=false/numInputBatches=64-24  28.1MB/s ± 3%   31.7MB/s ± 1%   +12.63%  (p=0.000 n=10+10)
Aggregator/SUM/hash/decimal/groupSize=1024/hasNulls=true/numInputBatches=64-24   29.9MB/s ± 1%   34.4MB/s ± 1%   +15.11%  (p=0.000 n=10+9)
Aggregator/SUM/hash/decimal/groupSize=2048/hasNulls=false/numInputBatches=64-24  28.8MB/s ± 0%   31.8MB/s ± 1%   +10.64%  (p=0.000 n=9+10)
Aggregator/SUM/hash/decimal/groupSize=2048/hasNulls=true/numInputBatches=64-24   30.6MB/s ± 0%   34.8MB/s ± 1%   +13.69%  (p=0.000 n=9+10)
Aggregator/SUM/hash/decimal/groupSize=4096/hasNulls=false/numInputBatches=64-24  29.1MB/s ± 1%   32.1MB/s ± 2%   +10.30%  (p=0.000 n=10+10)
Aggregator/SUM/hash/decimal/groupSize=4096/hasNulls=true/numInputBatches=64-24   30.9MB/s ± 1%   34.6MB/s ± 1%   +12.01%  (p=0.000 n=10+10)
```

asubiotto reviewed (commented):
> Reviewed 1 of 1 files at r1, 3 of 3 files at r2, 7 of 7 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47942)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng and @yuzefovich)

___
*[pkg/sql/colexec/hash_aggregator.go, line 195 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47942#-M5w474P8r1HP1r1jfNK:-M5w474P8r1HP1r1jfNL:b-ldkv4r) ([raw file](https://github.com/cockroachdb/cockroach/blob/4895de0c6cfb2de28a20ea45b4038cee2cbc75a5/pkg/sql/colexec/hash_aggregator.go#L195)):*
> ```Go
> 
> 	// We picked value this as the result of our benchmark.
> 	tupleLimit := coldata.BatchSize() * 2
> ```

This is interesting. Based on previous benchmarks, it seemed like it was good to have a buffering stage. What changed?
___
*[pkg/sql/colexec/hash_aggregator.go, line 374 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47942#-M5w3Ir12Xtyzum3KGjR:-M5w3Ir12Xtyzum3KGjS:bi8jpbp) ([raw file](https://github.com/cockroachdb/cockroach/blob/4895de0c6cfb2de28a20ea45b4038cee2cbc75a5/pkg/sql/colexec/hash_aggregator.go#L374)):*
> ```Go
> 	for selIdx, hashCode := range hashBuffer {
> 		selsSlot := -1
> 		for slot, hash := range op.scratch.hashCodeForSelsSlot {
> ```

maybe add a comment as to why we're not using a map
___
*[pkg/sql/colexec/hash_aggregator.go, line 447 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47942#-M5w4MMz06M8KhpXhioW:-M5w4MMz06M8KhpXhioX:bfgdqw) ([raw file](https://github.com/cockroachdb/cockroach/blob/69037788b1a10ca287d6b9b9b45b4e3c2cd870ba/pkg/sql/colexec/hash_aggregator.go#L447)):*
> ```Go
> }
> 
> const hashAggFuncsAllocSize = 128
> ```

What's the perf+mem difference with using 1024?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47942)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @Azhng)

___
*[pkg/sql/colexec/hash_aggregator.go, line 195 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47942#-M5w474P8r1HP1r1jfNK:-M5wkkHm0cGjB1bplk2i:b8nsr8n) ([raw file](https://github.com/cockroachdb/cockroach/blob/4895de0c6cfb2de28a20ea45b4038cee2cbc75a5/pkg/sql/colexec/hash_aggregator.go#L195)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

This is interesting. Based on previous benchmarks, it seemed like it was good to have a buffering stage. What changed?
</blockquote></details>

Not sure if it is the only reason, but the first commit here removes the lookup in a map (which has amortized `O(1)` cost) in favor of linear search in a slice (which has `O(distinct buffered tuples)` cost), so if we buffer up several batches, that cost will increase, especially in case of small group sizes.
___
*[pkg/sql/colexec/hash_aggregator.go, line 374 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47942#-M5w3Ir12Xtyzum3KGjR:-M5wncNE1Yy-t0_1QHs5:bqmjwcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/4895de0c6cfb2de28a20ea45b4038cee2cbc75a5/pkg/sql/colexec/hash_aggregator.go#L374)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

maybe add a comment as to why we're not using a map
</blockquote></details>

Added.
___
*[pkg/sql/colexec/hash_aggregator.go, line 447 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47942#-M5w4MMz06M8KhpXhioW:-M5wzRcrDmmpcn1FjEJr:bpuzfvs) ([raw file](https://github.com/cockroachdb/cockroach/blob/69037788b1a10ca287d6b9b9b45b4e3c2cd870ba/pkg/sql/colexec/hash_aggregator.go#L447)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

What's the perf+mem difference with using 1024?
</blockquote></details>

The comparison of 128 (old) vs 1024 (new) is [here](https://gist.github.com/yuzefovich/9263188a773fad6f113caaa169bd5ec9). Seems like there is no noticeable change in memory allocations, but there is a minor hit in performance with small group sizes (which is somewhat surprising to me).

I ran a few other comparisons: [128 vs 32](https://gist.github.com/yuzefovich/1ce2053301be878f9804144c61bc4e13) and [128 vs 64](https://gist.github.com/yuzefovich/ddf4084fb80ec8b5f9474f9d7f29d785)

Maybe 64 would be best? I feel like it is a nicer number than 128, but there is not much difference between the two in the benchmarks.


<!-- Sent from Reviewable.io -->

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 9 of 9 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47942#-:-M5zqteX6k9ujicwTNyU:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @Azhng, and @yuzefovich)

___
*[pkg/sql/colexec/hash_aggregator.go, line 447 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47942#-M5w4MMz06M8KhpXhioW:-M5zqkDt4DG_hNkglG8g:bgknw6h) ([raw file](https://github.com/cockroachdb/cockroach/blob/69037788b1a10ca287d6b9b9b45b4e3c2cd870ba/pkg/sql/colexec/hash_aggregator.go#L447)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

The comparison of 128 (old) vs 1024 (new) is [here](https://gist.github.com/yuzefovich/9263188a773fad6f113caaa169bd5ec9). Seems like there is no noticeable change in memory allocations, but there is a minor hit in performance with small group sizes (which is somewhat surprising to me).

I ran a few other comparisons: [128 vs 32](https://gist.github.com/yuzefovich/1ce2053301be878f9804144c61bc4e13) and [128 vs 64](https://gist.github.com/yuzefovich/ddf4084fb80ec8b5f9474f9d7f29d785)

Maybe 64 would be best? I feel like it is a nicer number than 128, but there is not much difference between the two in the benchmarks.
</blockquote></details>

Let's do 64 if there's not a big boost to using 128. Maybe also add a comment about how you got to this number.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47942#-:-M604RHf9AsX3JEOpC1g:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto and @Azhng)

___
*[pkg/sql/colexec/hash_aggregator.go, line 447 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47942#-M5w4MMz06M8KhpXhioW:-M604Jjb95aDZ7A3xjvi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/69037788b1a10ca287d6b9b9b45b4e3c2cd870ba/pkg/sql/colexec/hash_aggregator.go#L447)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Let's do 64 if there's not a big boost to using 128. Maybe also add a comment about how you got to this number.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->
